src/enclave/httpendpoint.h

```diff
@@ -133,6 +177,92 @@ namespace enclave
       flush();
     }
 
+    struct HandleProcessCbMsg
+    {
+      std::shared_ptr<Endpoint> self;
+      std::shared_ptr<JsonRpcContext> rpc_ctx;
+      std::shared_ptr<RpcHandler> search;
+    };
+
+    static void handle_process(
+      std::unique_ptr<enclave::Tmsg<HandleProcessCbMsg>> msg)
+    {
+      reinterpret_cast<HTTPServerEndpoint*>(msg->data.self.get())
+        ->handle_message_main_thread(msg->data.rpc_ctx, msg->data.search);
+    }
+
+    struct SendResponseVectCbMsg
+    {
+      std::vector<uint8_t> d = {};
+      http_status status;
+      std::shared_ptr<Endpoint> self;
+    };
+
+    static void send_response_vect(
+      std::unique_ptr<enclave::Tmsg<SendResponseVectCbMsg>> msg)
+    {
+      if (msg->data.d.size() != 0)
+      {
+        reinterpret_cast<HTTPServerEndpoint*>(msg->data.self.get())
+          ->send_response(msg->data.d);
+      }
+      else
+      {
+        reinterpret_cast<HTTPServerEndpoint*>(msg->data.self.get())
+          ->send_response(nullptr);
+      }
+    }
+
+    void handle_message_main_thread(
+      std::shared_ptr<JsonRpcContext>& rpc_ctx,
+      std::shared_ptr<RpcHandler>& search)
+    {
+      try
+      {
+        // we are executing here - send the thread 0 here
+        auto response = search->process(rpc_ctx);
+
+        if (!response.has_value())
+        {
+          // If the RPC is pending, hold the connection.
+          LOG_TRACE_FMT("Pending");
+          return;
+        }
+        else
+        {
+          // Otherwise, reply to the client synchronously.
+          LOG_TRACE_FMT("Responding");
+          auto msg = std::make_unique<enclave::Tmsg<SendResponseVectCbMsg>>(
+            &send_response_vect);
+          msg->data.status = HTTP_STATUS_OK;
+          msg->data.self = this->shared_from_this();
+          if (response.has_value())
+          {
+            msg->data.d = response.value();
+          }
+          else
+          {
+            msg->data.d.resize(0);
```

Comment:
If a RPC has been forwarded, I believe we can stop execution here and not return anything.

Reply:
Oh wait, we can never execute this code path because we already know that `response.has_value()` here.